### ISSUE TITLE
Do not run gopls if no open .go files

### ISSUE THREAD

**Issue by slayer** [microsoft/vscode-go#3042]
Hi

`gopls` is experimental project (It is currently in alpha, so it is not stable) and contains many memory leaks, on my laptop it can eat up to 4GB of RAM.

Is it possible to 
- do not run it if no `*.go` files are opened ?
- terminate it if no `*.go` files are opened ?
- periodically restart it ?

**Comment by hyangah** [comment]
"Go: Restart Language Server" from the command palette can be used to trigger language server restart. I think manual restart is better than periodically and automatically restarting it while interfering with users' go file editing.

Another option is to disable the language server for the workspace `"go.useLanguageServer": true" but that is not desirable if you are working in modules mode.

Since the `gopls` will be the default near future, it may be helpful to fix the memory leaks or problems sooner than later. Your help and bug reports are greatly appreciated. 

What version of `gopls` are you using? 
Can you tell more about the problem you are seeing and how to reproduce the issue?
@stamblerre

**Comment by stamblerre** [comment]
Thanks for the report, @slayer. I understand it can be frustrating when these types of bugs occur, but your suggestions will mask the problems, instead of helping us debug and fix them. `gopls` should not consume 4GB of RAM, so please [file an issue](https://github.com/golang/go/issues/new) with `gopls` if you see this. If possible, please attach your VS Code settings (`settings.json` file) and memory profiles. This can be done by adding the following settings to your VS Code `settings.json`:

```json
"go.languageServerFlags": [
    "-rpc.trace",
    "serve",
    "--debug=localhost:6060"
]
```

and running `go tool pprof -http :8080 http://localhost:6060/debug/pprof/heap`.

**Comment by slayer** [comment]
Thanks @hyangah 
I use gopls
```
golang.org/x/tools/gopls v0.3.1
    golang.org/x/tools/gopls@v0.3.1 h1:yNTWrf4gc4Or0UecjOas5pzOa3BL0WDDyKDV4Wz5VaM=
```

At the moment it is most RAM consuming process at the system (even bigger than chrome with slack and gmail :) )
```
 PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                   
 6794 vlad    20   0 1937.0m 850.1m  14.2m S   0.0   3.6   3:56.60 /home/vlad/go/bin/gopls -mode=stdio                                                       
```

> Can you tell more about the problem you are seeing and how to reproduce the issue?

Just editing `.go` files :)

my `settings.json`
```
...
    "go.useLanguageServer": true,
    "go.autocompleteUnimportedPackages": true,
    "go.formatTool": "goimports",
    "go.testOnSave": false,
    // "go.buildOnSave": "workspace",
    // "go.gocodeAutoBuild": true,
    "go.useCodeSnippetsOnFunctionSuggest": true,
    "go.lintTool": "golangci-lint"

    "go.languageServerExperimentalFeatures": {
        "format": true,
        "autoComplete": true
    },
    "gopls": {
        "usePlaceholders": true, // add parameter placeholders when completing a function
    },
    "[go]": {
        "editor.snippetSuggestions": "none",
        "editor.formatOnSave": true,
        "editor.codeActionsOnSave": {
            "source.organizeImports": true
        },
    },
    "go.lintTool": "golangci-lint"
...
```

@stamblerre thanks! will run it with profiling

**Comment by slayer** [comment]
I have uploaded goprof results to https://send.firefox.com/download/cb0f1a58824b60f8/#dL3s-VMmsJuu_043jKOvAg

**Comment by stamblerre** [comment]
Thank you for uploading that! As I suspected, `go/ssa` is using a lot of memory. This is because we recently added a new analyzer that uses `go/ssa`. We realized that this was an issue, so we have just released a new version of `gopls` (`gopls/v0.3.2`) that no longer has this analyzer enabled.

Can you try upgrading (VS Code should suggest it to you automatically, or you can run `GO111MODULE=on go get golang.org/x/tools/gopls@v0.3.2`) and see if this resolves the memory usage issue?
